src/typechecker/derive-type.lisp

```diff
@@ -460,6 +462,15 @@ EXPL-DECLARATIONS is a HASH-TABLE from SYMBOL to SCHEME"
                  (multiple-value-bind (typed-node binding-preds new-subs)
                      (derive-binding-type name tvar expr env subs name-map)
                    (setf subs new-subs)
+
+                   (with-type-context ("definition of ~A" name)
+                     (when (not allow-deferred-predicates)
```

Comment:
unless

Reply:
I think this is a bit easier to understand because it is not the default codepath